Mark volume as uncertain after Unmount* fails

jsafrane opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Mark volume device / volume mount as uncertain after UnmountDevice / UnmountVolume operation fails, because kubelet can't be sure what's the actual state of the volume. Kubelet will then call MountDevice / SetUp to make sure the volume is fully mounted when a new pod wants to use the same volume.

Similarly, update Unmap* calls for block volumes.

#### Which issue(s) this PR fixes:
Fixes: #100182

#### Special notes for your reviewer:

* Added e2e test for NodeUnstage errors, which exercises the UnmountDevice error path.
* Both transient and final errors result in the mount marked as uncertain. IMO it's not safe to assume that a mount is usable even after UnmountVolume / UnmountDevice final error. Better to call SetUp / MountDevice again when the mount is needed.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed starting new pods after previous pod timed out unmounting its volumes.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jsafrane committed:
Fix UnmountDevice error cases

When UnmountDevice fails, kubelet treat the volume mount as uncertain,
because it does not know at which stage UnmountDevice failed. It may be
already partially unmonted / destroyed.

As result, MountDevice will be performer when a new Pod is started on the
node after UnmountDevice faiure.

jsafrane committed:
Add e2e test for NodeUnstage error cases

yangjunmyfm192085 commented:
/test pull-kubernetes-unit

yangjunmyfm192085 commented:
/test pull-kubernetes-e2e-kind

jsafrane commented:
I tried to add e2e tests for the block mode, but the mock driver does not support block volumes.

jsafrane commented:
Ready for review (I tried to add e2e tests for block mode, but the mock driver does not support block)

cc @gnufied @msau42 @jingxu97 PTAL

jingxu97 commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -897,6 +897,128 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
+				name: "two pods: should call NodeStage after previous NodeUnstage final error",
+				expectedCalls: []csiCall{
+					{expectedMethod: "NodeStageVolume", expectedError: codes.OK},
+					{expectedMethod: "NodeUnstageVolume", expectedError: codes.InvalidArgument},
```
For InvalidArgument, do we need to mark as uncertain?

jsafrane replied:
It's an example of a "final" response, where we can be reasonably sure that e.g. provisioning is not in progress. Not sure if we can apply the same logic to Unstage.

jingxu97 commented:
/approve

jingxu97 commented:
PR lgtm. just want to make sure other reviewers have a chance to take a look.

jingxu97 commented:
/lgtm

gnufied commented:
Just planning to take a look today.

/hold

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1309,6 +1351,17 @@ func (og *operationGenerator) GenerateUnmapDeviceFunc(
+			eventErr, detailedErr := deviceToDetach.GenerateError("UnmapDevice.MarkDeviceAsUncertain failed", markDeviceUncertainErr)
+			return volumetypes.NewOperationContext(eventErr, detailedErr, migrated)
+		}
+
```
I see that we are marking device as uncertain even before unstage operation has started. This means that if `TearDownDevice` fails for a reason that is final (i.e which tells us unstage never really started) - then volume will still be treated as uncertain and hence a full stage has to be performed to make the volume usable again. 

This seems very different from how we handled uncertain devices/volumes in other places.

jingxu97 replied:
another way is perform teardown first. In case of error returned, mark it as uncertain. In case of no error, mark it as unmounted.

jsafrane replied:
I pointed out in the description of the PR, IMO it's not safe to assume that a mount is usable even after UnmountVolume / UnmountDevice final error. Better to call SetUp / MountDevice again when the mount is needed. It is a corner case already and one extra MountDevice / SetUp won't harm, especially if it's idempotent.

I am marking the volume as uncertain at the beginning, because unmapping the volume is a long process - see below - and it makes the error handling much easier.

gnufied replied:
I think my main concern is - other intree volume types do not benefit from being added to ASOW in uncertain state. Ideally all of `MountDevice` calls should be idempotent and hopefully nothing would break but it is somewhat of risky behaviour change and I am not sure if we have e2es that verify stuff like this. 

May be it is fine and may be - we will see some regressions.

jsafrane replied:
If any of the operation below fails (like `isDeviceOpened` check), the volume is still marked as mounted, while e.g. iscsi or FC volumes were already detached in `TearDownDevice`.  Therefore kubelet should mark it as uncertain. It is not related to CSI.

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1208,6 +1231,25 @@ func (og *operationGenerator) GenerateUnmapVolumeFunc(
+			VolumeMountState:    VolumeMountUncertain,
+		}
+		markVolumeUncertainErr := actualStateOfWorld.MarkVolumeMountAsUncertain(markVolumeOpts)
+		if markVolumeUncertainErr != nil {
```
I think this could mean - if a pod running on a node has one or more of volume failing to unmount, then pod can still be deleted because volume is being marked as `uncertain` and hence `GetMountedVolumesForPod` won't find uncertain volumes?

jingxu97 replied:
good point!

In current logic, pod will not be deleted though. GetMountedVolumesForPod will not find uncertain volumes, but it also check volumepaths from disk directly by calling getMountedVolumePathListFromDisk.

Now thinking this, for pod deletion, I am thinking a safer approach for GetMountedVolumesForPod should include uncertain state..?

gnufied replied:
> n current logic, pod will not be deleted though. GetMountedVolumesForPod will not find uncertain volumes, but it also check volumepaths from disk directly by calling getMountedVolumePathListFromDisk.

yeah I don't trust `getMountedVolumePathListFromDisk` completely. Also, what happens if driver created no mount points but `unpublish` still failed?

jsafrane replied:
Adding uncertain volumes to `GetMountedVolumesForPod` will run stats on them and could start pods. This needs to be smarter, perhaps extending only `podVolumesExist` to uncertain volumes.

gnufied replied:
I am trying to remember - in which cases we don't want an uncertain volume to be added in  `GetMountedVolumesForPod`? This would potentially prevent a pod from being deleted in some cases - but may be that is a good thing?

jsafrane replied:
When kubelet waits for all volumes to be mounted, it calls `verifyVolumesMountedFunc`, which calls `volumemanager.getUnmountedVolumes()`, which then calls  `GetMountedVolumesForPod`. If uncertain volumes are returned from `GetMountedVolumesForPod`, kubelet will interpret it as the volumes are mounted and it will start the pod.

jingxu97 replied:
we could have separate logic in these cases.
1. During pod start, when checking whether all volumes are already mounted, this check should not include uncertain volumes
2. During pod deletion, when checking whether volumes are unmounted, it should include uncertain volumes.

gnufied replied:
Yes I think that may work. I am thinking we should leave `GetMountedVolumesForPod` alone because it is also called from other places but may be introduce another function which would include uncertain and mounted volumes before allowing pod deletion.

jsafrane replied:
/lgtm cancel
This needs to be reworked as suggested above.

jsafrane replied:
I added `GetPossiblyMountedVolumesForPod` for the case when `podVolumesExist` needs to know when all volumes were really unmounted (and are not uncertain).
@gnufied @jingxu97, PTAL

ibotty commented:
Anything blocking it?

jsafrane committed:
Add GetPossiblyMountedVolumesForPod to let kubelet know all volumes were unmounted

podVolumesExist() should consider also uncertain volumes (where kubelet
does not know if a volume was fully unmounted) when checking for pod's
volumes. Added GetPossiblyMountedVolumesForPod for that.

Adding uncertain mounts to GetMountedVolumesForPod would potentially break
other callers (e.g. `verifyVolumesMountedFunc`).

ehashman commented:
/hold cancel

since I think the hold has been addressed... on the node reviewer backlog

jsafrane commented:
/hold
the test failures look genuine

```
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-link] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume [Slow]
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-link] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume [Slow]
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-link] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume [Slow]
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-link-bindmounted] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume 
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-link-bindmounted] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume [Slow]
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-link-bindmounted] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume [Slow]
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: tmpfs] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volumeKubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: tmpfs] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: tmpfs] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: blockfs] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: blockfs] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-bindmounted] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume 
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-bindmounted] [Testpattern: Pre-provisioned PV (block volmode)] volumeMode should fail to create pod by failing to mount volume [Slow]

jsafrane committed:
Add PodRemovedFromVolume

To know when a volume has been fully unmounted (incl. uncertain mounts).

jsafrane commented:
I added `PodRemovedFromVolume()` to ASW to have a way how DSW populator can see that a volume has been fully unmounted from a pod. `PodExistsInVolume`, which is used today, returns only fully mounted volumes and not the uncertain ones.

jsafrane commented:
/hold cancel

gnufied commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -278,12 +278,12 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
 		}
-		exists, _, _ := dswp.actualStateOfWorld.PodExistsInVolume(volumeToMount.PodName, volumeToMount.VolumeName)
+		removed := dswp.actualStateOfWorld.PodRemovedFromVolume(volumeToMount.PodName, volumeToMount.VolumeName)
 		var volumeToMountSpecName string
```
so if volume was uncertain, this would result in volume being deleted from DSOW. Previously we did exact opposite:

> if volume mount was uncertain we should keep trying to mount the volume

which means if volume-mgr started mounting the volume and say first mount failed and volume entered uncertain state, then we do not remove the volume from DSOW until volume is correctly mounted.  Now we would basically abort potentially in-flight mount request and issue umount requests.  I am not sure which one is more correct.

gnufied replied:
I think it is fine.

jsafrane replied:
> if volume mount was uncertain we should keep trying to mount the volume

We want to keep this behavior when a pod needs the volume. However, the change above is for deleted pods, where we want to "keep trying to **unmount** the volume". The mount may never succeed, because of timeouts or whatnot.

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -681,6 +696,31 @@ func (asw *actualStateOfWorld) PodExistsInVolume(
+		if podObj.volumeMountStateForPod == operationexecutor.VolumeMountUncertain {
+			return false
+		}
+		if podObj.volumeMountStateForPod == operationexecutor.VolumeMounted {
```
reading this again - so basically we are saying if a volume is uncertain or mounted - then it is *not* removed from pod in the kubelet and hence can be removed from DSOW.  

Previously we prevented removal of uncertain volumes from DSOW, but since we not stop deletion of pods with uncertain volumes, may be it is fine.

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -110,6 +110,16 @@ type ActualStateOfWorld interface {
+	// If a volume with the name volumeName does not exist in the list of
+	// attached volumes, true is returned indicating the given volume is not yet
+	// attached.
+	PodRemovedFromVolume(podName volumetypes.UniquePodName, volumeName v1.UniqueVolumeName) bool
```
typo in - "given pod does not exist"?

I would perhaps also add the context that - 

> If a volume is fully mounted or is in uncertain mount state for a pod, it is considered that pod still exists in volume manager's actual state of the world.

jsafrane replied:
Rephrased a bit, using your suggestion.

```
	// PodRemovedFromVolume returns true if the given pod does not exist in the list of
	// mountedPods for the given volume in the cache, indicating that the pod has
	// fully unmounted it or it was never mounted the volume.
	// If the volume is fully mounted or is in uncertain mount state for the pod, it is
	// considered that the pod still exists in volume manager's actual state of the world
	// and false is returned.
```

jingxu97 replied:
how about name the function PodVolumeNotExist?

jsafrane replied:
I think it would be confusing with `PodExistsInVolume` - these two methods are not simple negation of each other. I intentionally chose a different name.

gnufied commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -278,12 +278,12 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
 		}
-		if !exists && podExists {
+		if removed && podExists {
 			klog.V(4).InfoS("Actual state does not yet have volume mount information and pod still exists in pod manager, skip removing volume from desired state", "pod", klog.KObj(volumeToMount.Pod), "podUID", volumeToMount.Pod.UID, "volumeName", volumeToMountSpecName)
```
This is kind subtle behaviour change btw. Do we have tests for it? I think existing uncertain device and mount tests only cover setup or mount failures. Having said that - it might be tricky to test it, but it might be a good idea to have some coverage,

jsafrane replied:
Added an unit test. I had to refactor `findAndRemoveDeletedPods` tests a bit - there were two tests with identical initialization, I created a common function with it and added a third test.

jsafrane committed:
Refactor dswp unit tests

Change existing desiredStateOfWorldPopulator.findAndAddNewPods tests to use
a common initialization function.

jsafrane committed:
Add unit test for DSWP with uncertain volume

desiredStateOfWorldPopulator.findAndRemoveDeletedPods() should remove
volumes from DSW when a pod is deleted on the API server and the volume is
uncertain in ASW.

jingxu97 commented on pkg/kubelet/kubelet_volumes.go:
```diff
@@ -74,7 +74,7 @@ func (kl *Kubelet) ListBlockVolumesForPod(podUID types.UID) (map[string]volume.B
 // pods for the specified volume are mounted.
 func (kl *Kubelet) podVolumesExist(podUID types.UID) bool {
```
maybe also update the comments mentioning about uncertain volumes?

jsafrane replied:
Fixed

gnufied commented:
/retest

gnufied commented:
The test failure might be a flake.  This looks good to me. 

/lgtm

pacoxu commented:
/assign @random-liu
for approval from kubelet(sig-node)

gnufied commented:
/retest

jsafrane commented:
pull-kubernetes-e2e-gce-storage-slow won't pass, I am fixing it in https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/788

gnufied commented:
/assign @derekwaynecarr @ehashman

gnufied commented:
/retest
/lgtm

mrunalp commented:
/approve

timebertt commented:
Any chance we can backport this change?
We observed https://github.com/kubernetes/kubernetes/issues/100182 multiple times on v1.20 clusters.

ankitjain28may commented:
Any updates, can we get to know if this is getting backported to 1.20 since we are facing a lot of such issues causing production downtimes?
